Allow setting wmsMaxWidth and wmsMaxHeight in the server env

mbernasocchi opened the pull request:
## Description
This PR allows setting two env variable s to enforce MaxWidth and MaxHeight at the server level. 
The most conservative between the server env variable and the value defined in the project is used.

The use case in mind here is to protect a server that serves a lot of different projects from miss-configured or not configured projects.

The choice to use the most conservative setting reflects the idea of allowing the "project owner" to set a constraint but also to allow the server admin to maybe be more conservative depending on the server power.

Sponsored by Kt. Zug

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask i all your pointn a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

mbernasocchi committed:
Allow setting wmsMaxWidth and wmsMaxHeight in the server env

created env variables code

use server settings as well

add since 3.8

elpaso commented:
What's the use case? Why the limits set in the project are not enough?

mbernasocchi commented:
> What's the use case? Why the limits set in the project are not enough?

Hi @elpaso I updated the description

mbernasocchi committed:
Add test and sip

add missing methods

Signed-off-by: Marco Bernasocchi <marco@opengis.ch>

Update src/server/qgsserversettings.cpp

Co-Authored-By: mbernasocchi <marco@opengis.ch>

Update src/server/qgsserversettings.cpp

Co-Authored-By: mbernasocchi <marco@opengis.ch>

Update tests/src/python/test_qgsserver_wms_getmap.py

Co-Authored-By: mbernasocchi <marco@opengis.ch>

Update src/server/qgsserversettings.h

Co-Authored-By: mbernasocchi <marco@opengis.ch>

Adress most review comments in PR

Signed-off-by: Marco Bernasocchi <marco@opengis.ch>

m-kuhn commented on tests/src/python/test_qgsserver_wms_getmap.py:
```diff
@@ -617,6 +617,11 @@ def test_wms_getmap_background(self):
+
+        os.environ['QGIS_SERVER_WMS_MAX_WIDTH'] = 6000
+        os.environ['QGIS_SERVER_WMS_MAX_HEIGHT'] = 6000
+        # thest the 6000 limit from server is overriden by the more conservative 5000 in the project
```
```suggestion
        # test the 6000 limit from server is overriden by the more conservative 5000 in the project
```

m-kuhn commented on src/server/qgsserversettings.cpp:
```diff
@@ -162,6 +162,27 @@ void QgsServerSettings::initSettings()
+  // max height
+  const Setting sMaxHeight = { QgsServerSettingsEnv::QGIS_SERVER_WMS_MAX_HEIGHT,
+                               QgsServerSettingsEnv::DEFAULT_VALUE,
+                               "Maximum height for a WMS request. The most conservative between this and the project one is used",
```
```suggestion
                               "Maximum height for a WMS request. The lower one of this and the project configuration is used.",
```

I think this would be a bit clearer / easier to read?

m-kuhn commented on src/server/qgsserversettings.h:
```diff
@@ -183,6 +185,20 @@ class SERVER_EXPORT QgsServerSettings
+  
+    /**
+     * Returns the max height of a WMS GetMap request.
+     * \returns Returns the max height of a WMS GetMap request.
```
```suggestion
     * \returns the max height of a WMS GetMap request.
```

m-kuhn commented on src/server/qgsserversettings.h:
```diff
@@ -183,6 +185,20 @@ class SERVER_EXPORT QgsServerSettings
     bool showGroupSeparator() const;
+  
+    /**
+     * Returns the max height of a WMS GetMap request.
```
Maybe also add a comment here that this acts as a limit for any project configuration?

pblottiere replied:
Due to the recent work of @signedav in https://github.com/qgis/QGIS/pull/9545, it'll also have an impact on `GetLegendGraphics` request. I don't know if it's wanted, but in this case, it should be stated in the documentation too.

mbernasocchi replied:
@pblottiere I think it is fair to use it also on the GetLegendGraphics, since the settings in the project would also impact that. no?

pblottiere replied:
Hi @mbernasocchi,

> I think it is fair to use it also on the GetLegendGraphics, since the settings in the project would also impact that. no?

Good point :).

And thanks for your work!

andreasneumann commented:
Once this PR is ready, I'd like to see backports to 3.4 (LTR) and 3.6.

It is an important security feature that protects your QGIS server from dying, because people out there can request huge image sizes and I can tell that sometimes people (at least in our organization) sometimes forget to set max_width/height in the OWS server tab in the projects.

I know, one can also set this on the project level - but in a larger organization - if you have many projects - there could always be a project that doesn't properly set max_width and max_height correctly.

@pblottiere @rldhont what do you think?

elpaso commented on src/server/qgsserversettings.cpp:
```diff
@@ -162,6 +162,27 @@ void QgsServerSettings::initSettings()
+  // max height
+  const Setting sMaxHeight = { QgsServerSettingsEnv::QGIS_SERVER_WMS_MAX_HEIGHT,
+                               QgsServerSettingsEnv::DEFAULT_VALUE,
+                               "Maximum height for a WMS request. The most conservative between this and the project one is used",
```
@m-kuhn why no `QStringLiteral` is required here? (and below)

m-kuhn replied:
Hmmm... I didn't pay attention to that, it would probably be good?

mbernasocchi replied:
@m-kuhn I dit the same as in https://github.com/qgis/QGIS/blob/0bc43b6dc7f8a514b164277c133a138afdd04533/src/server/qgsserversettings.cpp#L124 and all other Settings
do you want me to add QStringLiteral to all of them?

m-kuhn replied:
Changing this everywhere sounds good to me

pblottiere replied:
> Changing this everywhere sounds good to me

+1

elpaso commented on src/server/qgsserversettings.cpp:
```diff
@@ -353,3 +374,12 @@ bool QgsServerSettings::showGroupSeparator() const
 
+int QgsServerSettings::wmsMaxHeight() const
+{
+  return value( QgsServerSettingsEnv::QGIS_SERVER_WMS_MAX_WIDTH ).toInt();
```
```suggestion
  return value( QgsServerSettingsEnv::QGIS_SERVER_WMS_MAX_HEIGHT ).toInt();
```

pblottiere commented on src/server/services/wms/qgswmsrenderer.cpp:
```diff
@@ -1839,21 +1840,51 @@ namespace QgsWms
+    if ( wmsMaxWidthEnv != -1 && wmsMaxWidthProj != -1 )
+    {
+      // both are set, so we take the more conservative one
+      wmsMaxHeight = qMin( wmsMaxHeightProj, wmsMaxHeightEnv );
```
`qMin` -> `std::min`

pblottiere commented on src/server/services/wms/qgswmsrenderer.cpp:
```diff
@@ -1839,21 +1840,51 @@ namespace QgsWms
+    else
+    {
+      // none or one are set, so we take the bigger one which is the one set or -1
+      wmsMaxHeight = qMax( wmsMaxHeightProj, wmsMaxHeightEnv );
```
`qMax` -> `std::max`

pblottiere commented on src/server/services/wms/qgswmsrenderer.cpp:
```diff
@@ -1839,21 +1840,51 @@ namespace QgsWms
+    if ( wmsMaxWidthEnv != -1 && wmsMaxWidthProj != -1 )
+    {
+      // both are set, so we take the more conservative one
+      wmsMaxWidth = qMin( wmsMaxWidthProj, wmsMaxWidthEnv );
```
`qMin` -> `std::min`

pblottiere commented on src/server/services/wms/qgswmsrenderer.cpp:
```diff
@@ -1839,21 +1840,51 @@ namespace QgsWms
+    else
+    {
+      // none or one are set, so we take the bigger one which is the one set or -1
+      wmsMaxWidth = qMax( wmsMaxWidthProj, wmsMaxWidthEnv );
```
`qMax` -> `std::max`

pblottiere commented on src/server/services/wms/qgswmsrenderer.h:
```diff
@@ -283,6 +283,8 @@ namespace QgsWms
 
       QgsWmsParameters mWmsParameters;
 
+      QgsServerSettings mServerSettings;
```
You don't need to add a member, it's already available through `mContext`.

pblottiere commented on src/server/services/wms/qgswmsrenderer.cpp:
```diff
@@ -1839,21 +1840,51 @@ namespace QgsWms
-    int wmsMaxHeight = QgsServerProjectUtils::wmsMaxHeight( *mProject );
+    //HEIGHT
+    int wmsMaxHeightProj = QgsServerProjectUtils::wmsMaxHeight( *mProject );
+    int wmsMaxHeightEnv = mServerSettings.wmsMaxHeight();
```
`mContext.settings().wmsMaxHeight()`

pblottiere commented on src/server/services/wms/qgswmsrenderer.cpp:
```diff
@@ -1839,21 +1840,51 @@ namespace QgsWms
+    //and WIDTH / HEIGHT parameter is in the range allowed range
+    //WIDTH
+    int wmsMaxWidthProj = QgsServerProjectUtils::wmsMaxWidth( *mProject );
+    int wmsMaxWidthEnv = mServerSettings.wmsMaxWidth();
```
`mContext.settings().wmsMaxWidth()`

pblottiere commented:
Hi @andreasneumann,

> I know, one can also set this on the project level - but in a larger organization - if you have many projects - there could always be a project that doesn't properly set max_width and max_height correctly.

> @pblottiere @rldhont what do you think?

I think it's a good safeguard to have :+1:.

mbernasocchi committed:
use QStringLiteral in Settings

Signed-off-by: Marco Bernasocchi <marco@opengis.ch>

remove unused mServerSettings

Signed-off-by: Marco Bernasocchi <marco@opengis.ch>

Fix typo and code style

Signed-off-by: Marco Bernasocchi <marco@opengis.ch>

Fix test

Fix test

mbernasocchi commented:
@elpaso, @pblottiere, @m-kuhn thanks for all the feedback, I addressed all your point

m-kuhn commented:
Leaving the decision to @pblottiere and @rldhont , please label it accordingly.

I'd say as a first step you make this ready for review.

rldhont commented:
Like @pblottiere I think it's a good idea to backport it.

m-kuhn reviewed (approved):
Ok, I've done all the admin on this PR, let's wait for travis

mbernasocchi commented:
great

m-kuhn commented:
Next time: `s/Fix test/Fix typo and code style/` ;)
Thanks!

mbernasocchi commented:
Done :)

m-kuhn commented:
@mbernasocchi I did a "fix test" commit for you

mbernasocchi commented:
> @mbernasocchi I did a "fix test" commit for you

Thanks a lot, but I know your master plan is to steal my commits ;)

m-kuhn commented:
The next one is on you https://travis-ci.org/qgis/QGIS/jobs/515579728#L2245-L2252

mbernasocchi commented:
yes already looking

elpaso commented:
@mbernasocchi I wonder if it would be wise to disclose the limits in the error message:

```
InvalidParameterValue">The requested map size is too large (maximum allowed: 11111 x 22222)
```

If not, at least in the server logs (if not yet done).

mbernasocchi committed:
Move invalid size test to its own class so it can reread os.env

Signed-off-by: Marco Bernasocchi <marco@opengis.ch>

add init_server method

Signed-off-by: Marco Bernasocchi <marco@opengis.ch>

Use own qgs application so we can change the env settings

Signed-off-by: Marco Bernasocchi <marco@opengis.ch>

mbernasocchi committed:
Separate the tests in two files since server uses static initialization
and thus we cannot change it's settings once it is created

Signed-off-by: Marco Bernasocchi <marco@opengis.ch>

Signed-off-by: Marco Bernasocchi <marco@opengis.ch>

m-kuhn commented on tests/src/python/test_qgsserver_wms_getmap_size_project.py:
```diff
@@ -0,0 +1,73 @@
+# -*- coding: utf-8 -*-
+"""QGIS Unit tests for QgsServer MaxHeight and MaxWidth Override Options.
+
+From build dir, run: ctest -R PyQgsServerGetMapSize -V
```
Leftover?

m-kuhn replied:
Ah no, sorry
